crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -591,6 +629,75 @@ mod tests {
         assert_eq!(i1.pos_vec(&db), &[ta, t1]);
     }
 
+    #[test]
+    fn intersection_negation_distributes_over_union() {
+        let db = setup_db();
+        let st = KnownClass::Sized.to_instance(&db);
+        let ht = KnownClass::Hashable.to_instance(&db);
+        // sh_t: Sized & Hashable
+        let sh_t = IntersectionBuilder::new(&db)
+            .add_positive(st)
+            .add_positive(ht)
+            .build()
+            .expect_intersection();
+        assert_eq!(sh_t.pos_vec(&db), &[st, ht]);
+        assert_eq!(sh_t.neg_vec(&db), &[]);
+
+        // ~sh_t => ~Sized | ~Hashable
+        let not_s_h_t = IntersectionBuilder::new(&db)
+            .add_negative(Type::Intersection(sh_t))
+            .build()
+            .expect_union();
+
+        // should have as elements: (~Sized),(~Hashable)
+        let not_st = st.negate(&db);
+        let not_ht = ht.negate(&db);
+        assert_eq!(not_s_h_t.elements(&db), &[not_st, not_ht]);
+    }
+
+    #[test]
+    fn mixed_intersection_negation_distributes_over_union() {
+        let db = setup_db();
+        let it = KnownClass::Int.to_instance(&db);
+        let st = KnownClass::Sized.to_instance(&db);
+        let ht = KnownClass::Hashable.to_instance(&db);
+        // sh_t: Sized & ~Hashable
+        let s_not_h_t = IntersectionBuilder::new(&db)
+            .add_positive(st)
+            .add_negative(ht)
+            .build()
+            .expect_intersection();
+        assert_eq!(s_not_h_t.pos_vec(&db), &[st]);
+        assert_eq!(s_not_h_t.neg_vec(&db), &[ht]);
+
+        // let's build Int & ~(Sized & ~Hashable)
+        let tt = IntersectionBuilder::new(&db)
+            .add_positive(it)
+            .add_negative(Type::Intersection(s_not_h_t))
+            .build();
+
+        // Int & ~(Sized & ~Hashable)
+        // -> Int & (~Sized | Hashable)
+        // -> (Int & ~Sized) | (Int & Hashable)
+        let tt = tt.expect_union();
+
+        assert_eq!(
+            tt.elements(&db),
+            &[
+                // Int & ~Sized
+                IntersectionBuilder::new(&db)
+                    .add_positive(it)
+                    .add_negative(st)
+                    .build(),
```

Comment:
a part of me wants something like `Type::intersection(elts: Vec<Type<'db>>)` for testing but I got pretty used to writing out the intersection builder at this point 😆

Reply:
Ha, you aren't the first to say this after working on IntersectionBuilder, cc @sharkdp 😆 It's not a problem at all to add test helpers to make tests less verbose, someone just needs to feel the pain sufficiently to actually do it!